src/term/transform/float_combinators.rs

```diff
@@ -1,277 +1,151 @@
+  name_gen: &mut usize,
+  term: &mut Term,
+  builtin: bool,
+  combinators: &mut BTreeMap<Name, Definition>,
+) {
+  let comb_name = Name::new(format!("{}$C{}", def_name, *name_gen));
+  *name_gen += 1;
 
-      (Detach::Combinator, Detach::Recursive) => Detach::Recursive,
-      (Detach::Recursive, Detach::Combinator) => Detach::Recursive,
-      (Detach::Recursive, Detach::Recursive) => Detach::Recursive,
+  let comb_var = Term::Ref { nam: comb_name.clone() };
+  let extracted_term = std::mem::replace(term, comb_var);
 
-      // TODO: Is this the best way to best deal with a term that is both unscoped and recursive?
-      (unscoped @ Detach::Unscoped { .. }, Detach::Recursive) => unscoped,
-      (Detach::Recursive, unscoped @ Detach::Unscoped { .. }) => unscoped,
-    }
-  }
+  let rules = vec![Rule { body: extracted_term, pats: Vec::new() }];
+  let rule = Definition { name: comb_name.clone(), rules, builtin };
+  combinators.insert(comb_name, rule);
 }
 
 impl Term {
-  pub fn float_combinators(&mut self, def_name: &Name, builtin: bool, combinators: &mut Combinators) {
-    self.go_float(0, &mut TermInfo::new(def_name.clone(), builtin, combinators));
-  }
-
-  fn go_float(&mut self, depth: usize, term_info: &mut TermInfo) -> Detach {
+  /// A term can be considered safe if it is:
+  /// - A Number or an Eraser.
+  /// - A Tuple or Superposition where all elements are safe.
+  /// - A constant Lambda, e.g. a nullary constructor.
+  /// - A Reference with safe body.
+  pub fn is_safe(&self, book: &Book, seen: &mut IndexSet<Name>) -> bool {
     Term::recursive_call(move || match self {
-      Term::Lam { .. } | Term::Chn { .. } if self.is_id() => Detach::Combinator,
+      Term::Num { .. } | Term::Era => true,
 
-      Term::Lam { .. } => self.float_lam(depth, term_info),
-      Term::Chn { .. } => self.float_lam(depth, term_info),
-      Term::App { .. } => self.float_app(depth, term_info),
-      Term::Mat { .. } => self.float_mat(depth, term_info),
+      Term::Tup { els } | Term::Sup { els, .. } => els.iter().all(|e| Term::is_safe(e, book, seen)),
 
-      Term::Var { nam } => {
-        term_info.request_name(nam);
-        Detach::Combinator
-      }
+      Term::Lam { .. } => self.is_constant_lambda(book, seen),
 
-      Term::Lnk { nam } => Detach::unscoped_var(nam.clone()),
-
-      Term::Ref { nam: def_name } if def_name == &term_info.def_name => Detach::Recursive,
+      Term::Ref { nam } => {
+        if seen.contains(nam) {
+          return false;
+        }
 
-      _ => {
-        let mut detach = Detach::Combinator;
-        for (child, binds) in self.children_mut_with_binds() {
-          detach = detach & child.go_float(depth + 1, term_info);
-          for bind in binds {
-            term_info.provide(bind.as_ref());
-          }
+        seen.insert(nam.clone());
+        if let Some(definition) = book.defs.get(nam) {
+          definition.rule().body.is_safe(book, seen)
+        } else {
+          false
```

Comment:
I think we'll never find a reference to one of the newly created definitions, so even throwing an error would be fine. But this is also ok